modules/framework/galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/rbac/CacheRBACImpl.java

```diff
@@ -17,37 +21,72 @@
 
 public class CacheRBACImpl implements CacheRBAC {
 
+    // Only keep users-to-actions entries in the cache for 24 hours
+    private static final long CACHED_ACTIONS_TIME_TO_LIVE_SECS = 24 * 60 * 60;
+
+    private static final String ACTIONS_PROPERTY_SUFFIX = ".actions";
+
+    private IDynamicStatusStoreService dssService;
     private IAuthStoreService authStoreService;
     private RBACService rbacService;
-    
-    public CacheRBACImpl(IAuthStoreService authStoreService, RBACService rbacService) {
+
+    public CacheRBACImpl(
+        IDynamicStatusStoreService dssService,
+        IAuthStoreService authStoreService,
+        RBACService rbacService
+    ) {
+        this.dssService = dssService;
         this.authStoreService = authStoreService;
         this.rbacService = rbacService;
     }
 
     @Override
     public synchronized void addUser(String loginId, List<String> actionIds) throws RBACException {
-        // Do nothing for now...
+        try {
+            // Create a property in the DSS of the form:
+            // dss.rbac.<loginId>.actions = <comma-separated action IDs>
+            String commaSeparatedActionIds = String.join(",", actionIds);
+            String actionsKey = loginId + ACTIONS_PROPERTY_SUFFIX;
+            dssService.put(actionsKey, commaSeparatedActionIds, CACHED_ACTIONS_TIME_TO_LIVE_SECS);
+        } catch (DynamicStatusStoreException e) {
+            throw new RBACException("Failed to cache user actions", e);
+        }
     }
 
     @Override
     public synchronized boolean isActionPermitted(String loginId, String actionId) throws RBACException {
-        IUser user = getUserFromAuthStore(loginId);
-        String userRoleId = user.getRoleId();
-        if (userRoleId == null) {
-            userRoleId = rbacService.getDefaultRoleId();
-            user.setRoleId(userRoleId);
-        }
-        Role userRole = rbacService.getRoleById(userRoleId);
+        boolean isActionPermitted = false;
+        try {
+            String userActionsKey = loginId + ACTIONS_PROPERTY_SUFFIX;
+            String commaSeparatedUserActions = dssService.get(userActionsKey);
+
+            List<String> userActions = new ArrayList<>();
```

Comment:
Use Set<String> and HashSet<String> as the contains() lookup is loads quicker.

Reply:
Done in https://github.com/galasa-dev/galasa/pull/162/commits/011d27b7bf48c4812b532fdbf9eb7eb2c9a7be3c